cmd/object-handlers.go

```diff
@@ -3265,24 +3265,41 @@ func (api objectAPIHandlers) PutObjectTaggingHandler(w http.ResponseWriter, r *h
 		opts.UserDefined[xhttp.AmzBucketReplicationStatus] = replication.Pending.String()
 	}
 
+	objInfo, err := objAPI.GetObjectInfo(ctx, bucket, object, opts)
+	if err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
+
+	tagsStr := tags.String()
+
 	// Put object tags
-	err = objAPI.PutObjectTags(ctx, bucket, object, tags.String(), opts)
+	err = objAPI.PutObjectTags(ctx, bucket, object, tagsStr, opts)
 	if err != nil {
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
 
 	if replicate {
```

Comment:
why are we doing an extra call here? objInfo is only used if `replication is enabled - doing it again and again for every PutObjectTags is an expensive operation..

Reply:
> why are we doing an extra call here? objInfo is only used if `replication is enabled - doing it again and again for every PutObjectTags is an expensive operation..

Are you asking why we have GetObjectInfo here ? the reason is that mc mirror needs to know the version-id of the object that we are setting the tag

Reply:
Yes but this is an incorrect change, PutObjectTags should return object info for the version it modified.

Making double calls is not useful here.

Reply:
> Making double calls is not useful here.

oh okay you are saying that we should optimize PutObjectTags API, yeah will do